### ISSUE TITLE
ESP32-CAM sned picture. Half of it arrives ok.

### ISSUE THREAD

**Issue by pcbtborges** [mobizt/ESP32-Mail-Client#28]
Hi, I am trying to send pictures via gmail.
Only part of the image arrives (such as the top half): smtpData.addAttachData("myPic.jpg", "image/jpg", fb->buf, fb->len);
Nothing arrives: smtpData.addAttachFile(path.c_str());
Arduino IDE 1.8.10
Mail Client Arduino Library for ESP32 v 2.0.4

`
#define DEBUG_ESP    //comment out to activate debug console

#ifdef DEBUG_ESP
  #define DBG(x)   Serial.print(x)
  #define DBGln(x) Serial.println(x)
#else 
  #define DBG(...)
  #define DBGln(...)
#endif

#include <WiFiClientSecure.h>
#include <NTPClient.h>
#include <WiFiUdp.h>
WiFiUDP ntpUDP;
NTPClient timeClient(ntpUDP);

#include "ESP32_MailClient.h"
//The Email Sending data object contains config and data to send
SMTPData smtpData;

// Your WiFi credentials.
char ssid[] = "mySSID";
char pass[] = "myPASS";

#include "esp_camera.h"
#include "esp_timer.h"
#include "img_converters.h"
#include "Arduino.h"
#include "fb_gfx.h"
#include "fd_forward.h"
#include "fr_forward.h"
#include "FS.h"                // SD Card ESP32
#include "SD_MMC.h"            // SD Card ESP32
#include "soc/soc.h"           // Disable brownour problems
#include "soc/rtc_cntl_reg.h"  // Disable brownour problems
//#include "dl_lib.h"
#include "driver/rtc_io.h"

RTC_DATA_ATTR int bootCount = 0;

// Pin definition for CAMERA_MODEL_AI_THINKER
#define PWDN_GPIO_NUM     32
#define RESET_GPIO_NUM    -1
#define XCLK_GPIO_NUM      0
#define SIOD_GPIO_NUM     26
#define SIOC_GPIO_NUM     27
#define Y9_GPIO_NUM       35
#define Y8_GPIO_NUM       34
#define Y7_GPIO_NUM       39
#define Y6_GPIO_NUM       36
#define Y5_GPIO_NUM       21
#define Y4_GPIO_NUM       19
#define Y3_GPIO_NUM       18
#define Y2_GPIO_NUM        5
#define VSYNC_GPIO_NUM    25
#define HREF_GPIO_NUM     23
#define PCLK_GPIO_NUM     22
 
String pictureName;
String path;
int shots=5;

#define uS_TO_S_FACTOR 1000000

//Callback function to get the Email sending status
void sendCallback(SendStatus info);

//init camera buffer
camera_fb_t * fb PROGMEM = NULL;
  
void setup() {
  WRITE_PERI_REG(RTC_CNTL_BROWN_OUT_REG, 0); //disable brownout detector
  Serial.begin(115200);
  Serial.setDebugOutput(true);

  Serial.println("Starting SD Card...");
  if( !SD_MMC.begin( "/sdcard", true ) ) { // slow 1bit mode
    DBGln( "SD card init failed" );
    return;
  }  
  uint8_t cardType = SD_MMC.cardType();
  if(cardType == CARD_NONE){
    Serial.println("No SD Card attached");
    return;
  }

  while (1) {
    activateWiFi();
    if(WiFi.status() == WL_CONNECTED){
      break; 
    } else {
      DBGln("Aguardando 15min para tentar WiFi de novo...");
      delay(900000);
    }
  }  
 
  camera_config_t config;
  config.ledc_channel = LEDC_CHANNEL_0;
  config.ledc_timer = LEDC_TIMER_0;
  config.pin_d0 = Y2_GPIO_NUM;
  config.pin_d1 = Y3_GPIO_NUM;
  config.pin_d2 = Y4_GPIO_NUM;
  config.pin_d3 = Y5_GPIO_NUM;
  config.pin_d4 = Y6_GPIO_NUM;
  config.pin_d5 = Y7_GPIO_NUM;
  config.pin_d6 = Y8_GPIO_NUM;
  config.pin_d7 = Y9_GPIO_NUM;
  config.pin_xclk = XCLK_GPIO_NUM;
  config.pin_pclk = PCLK_GPIO_NUM;
  config.pin_vsync = VSYNC_GPIO_NUM;
  config.pin_href = HREF_GPIO_NUM;
  config.pin_sscb_sda = SIOD_GPIO_NUM;
  config.pin_sscb_scl = SIOC_GPIO_NUM;
  config.pin_pwdn = PWDN_GPIO_NUM;
  config.pin_reset = RESET_GPIO_NUM;
  config.xclk_freq_hz = 20000000;
  config.pixel_format = PIXFORMAT_JPEG;
   
  if(psramFound()){
    config.frame_size = FRAMESIZE_UXGA; // FRAMESIZE_ + QVGA|CIF|VGA|SVGA|XGA|SXGA|UXGA
    config.jpeg_quality = 10;
    config.fb_count = 2;
  } else {
    config.frame_size = FRAMESIZE_SVGA;
    config.jpeg_quality = 12;
    config.fb_count = 1;
  }
 
  // Init Camera
  esp_err_t err = esp_camera_init(&config);
  if (err != ESP_OK) {
    Serial.printf("Camera init failed with error 0x%x", err);
    return;
  }
  rtc_gpio_hold_dis(GPIO_NUM_4);
  pinMode(4, OUTPUT);
  digitalWrite(4, LOW);  
  for(int i=0;i<shots;i++){ //takes x pics
    
    // Take Picture with Camera
    digitalWrite(4, HIGH); 
    delay(30);   
    fb = esp_camera_fb_get();  
    digitalWrite(4, LOW);  

    if(!fb) {
      Serial.println("Camera capture failed");
      break;
    }
    String pic_name = getDateTime();
    //pictureName = pic_name.substring(0,2)+ pic_name.substring(6);
 
    // Path where new picture will be saved in SD Card
    path = "/" + pic_name +".jpg";
    pictureName = pic_name;
 
    fs::FS &fs = SD_MMC;
    //Serial.printf("Picture file name: %s\n", path.c_str());
 
    File file = fs.open(path.c_str(), FILE_WRITE);
    if(!file){
      Serial.println("Failed to open file in writing mode");
    } else {
      file.write(fb->buf, fb->len); // payload (image), payload length
      Serial.printf("Saved picture to path: %s\n", path.c_str());
    }
    file.close();
    smtpMail(path);
    esp_camera_fb_return(fb);  //clear cam memory buffer

    delay(3000);
  }
} 
 
void loop() { 
}

String getDateTime(){
  String curDateTime="";
  String formattedDate;
  String dayStamp;
  String timeStamp;

  while(!timeClient.update()) {
    timeClient.forceUpdate();
  }
  // The formattedDate comes with the following format:
  // 2018-05-28T16:00:13Z
  // We need to extract date and time
  formattedDate = timeClient.getFormattedDate();

  #ifdef DEBUG_ESP
    // Extract date
    int splitT = formattedDate.indexOf("T");
    dayStamp = formattedDate.substring(0, splitT);
    // Extract time
    timeStamp = formattedDate.substring(splitT+1, formattedDate.length()-1);
    Serial.print(formattedDate);  
    Serial.print(", Date: ");
    Serial.print(dayStamp);
    Serial.print(" - Time: ");
    Serial.println(timeStamp);    
  #endif 
  
  curDateTime  = formattedDate.substring(8,10);  //day
  curDateTime += formattedDate.substring(5,7);   //month
  curDateTime += formattedDate.substring(2,4);   //year
  curDateTime += formattedDate.substring(11,13); //hour
  curDateTime += formattedDate.substring(14,16); //minute
  curDateTime += formattedDate.substring(17,19); //second
  return curDateTime;
}  

void activateWiFi() {
  long locTimer;
  byte hardwareMAC[6];  //WIFI MAC address
  int counter = 0;
  //Activate and connect to WiFi 
  
  WiFi.disconnect(true);                     // disconnects STA Mode
  WiFi.mode(WIFI_STA);  
  WiFi.setSleep(false);  
  WiFi.begin(ssid, pass);
   
  DBG("Conectando com ");
  DBG(ssid);
  while (WiFi.status() != WL_CONNECTED) {
    DBG(".");
    locTimer = millis();
    while(locTimer+500>millis()){}
    counter++;
    if (counter > 15) {
      break;
    }
  }
  DBGln("");
  if (WiFi.status() == WL_CONNECTED) {
    DBG("Conectado com sucesso à rede WIFI ");
    DBGln(ssid);
    DBG("IP: ");   
    DBG(WiFi.localIP());    
    DBG(" - ");
    WiFi.macAddress(hardwareMAC);
    DBG("Hard MAC: ");
    Serial.printf("%02x:%02x:%02x:%02x:%02x:%02x\n",hardwareMAC[5], hardwareMAC[4], hardwareMAC[3], hardwareMAC[2], hardwareMAC[1], hardwareMAC[0]);         
  } else {
    DBG("Timeout, verifique credenciais WiFi: ");
    DBGln(ssid);
  }
} 

//To use send Email for Gmail to port 465 (SSL), less secure app option should be enabled. https://myaccount.google.com/lesssecureapps?pli=1

//To receive Email for Gmail, IMAP option should be enabled. https://support.google.com/mail/answer/7126229?hl=en

void smtpMail(String path){

  //Set the Email host, port, account and password
  smtpData.setLogin("smtp.gmail.com", 465, "myACCT@gmail.com", "myMailPass"); //587 more secure

  //For library version 1.2.0 and later which STARTTLS protocol was supported,the STARTTLS will be 
  //enabled automatically when port 587 was used, or enable it manually using setSTARTTLS function.
  //smtpData.setSTARTTLS(true);

  //Set the sender name and Email
  smtpData.setSender("ESP32", "myGMail@gmail.com");

  //Set Email priority or importance High, Normal, Low or 1 to 5 (1 is highest)
  smtpData.setPriority("High");

  //Set the subject
  smtpData.setSubject("ESP32 SMTP Mail Sending Test");

  //Set the message - normal text or html format
  smtpData.setMessage("<div style=\"color:#ff0000;font-size:20px;\">Hello World! - From ESP32</div>", true);

  //Add recipients, can add more than one recipient
  smtpData.addRecipient("myYahooAcct@yahoo.com");

  //Add attachments, can add the file or binary data from flash memory, file in SD card
  //Data from internal memory
  //smtpData.addAttachData("firebase_logo.png", "image/png", (uint8_t *)dummyImageData, sizeof dummyImageData);
  smtpData.addAttachData("myPic.jpg", "image/jpg", fb->buf, fb->len);
  
  //Add attach files from SD card
  //Comment these two lines, if no SD card connected  
  //Two files that previousely created.
  //smtpData.addAttachFile("/text_file.txt");
  //smtpData.addAttachFile("/binary_file.dat");
  smtpData.addAttachFile(path.c_str());

  //Add some custom header to message
  //See https://tools.ietf.org/html/rfc822
  //These header fields can be read from raw or source of message when it received)
  smtpData.addCustomMessageHeader("Date: Sat, 10 Aug 2019 21:39:56 -0700 (PDT)");
  //Be careful when set Message-ID, it should be unique, otherwise message will not store
  //smtpData.addCustomMessageHeader("Message-ID: <abcde.fghij@gmail.com>");

  //Set the storage types to read the attach files (SD is default)
  //smtpData.setFileStorageType(MailClientStorageType::SPIFFS);
  smtpData.setFileStorageType(MailClientStorageType::SD);

  smtpData.setSendCallback(sendCallback);

  Serial.println("Sending mail now...");

  //Start sending Email, can be set callback function to track the status
  if (!MailClient.sendMail(smtpData)){
    Serial.println("Error sending Email, " + MailClient.smtpErrorReason());
  }

  //Clear all data from Email object to free memory
  smtpData.empty();
}

//Callback function to get the Email sending status
void sendCallback(SendStatus msg){
  //Print the current status
  Serial.println(msg.info());

  //Do something when complete
  if (msg.success())  {
    Serial.println("----------------");
  }
}

`

Also tried pictures with with 8.3 naming format as 12345678.jpg and pic names starting by char instead of numbers like: a1234567.jpg but attachments are still not sent.

Thanks in advance.

**Comment by mobizt** [comment]
The library was not support SD_MMC library, only SD.

You not need to use SD_MMC library to control and access only SD card (no MMC medias).

You can use SD.h instead, see [this configuration](https://github.com/mobizt/ESP32-Mail-Client/issues/23).

The function **addAttachData** required uint8_t array and its length, make sure that you have enough memory and valid array length.

The image data for **addAttachData** should be saved to flash (SPIFFS) instead of ram to save your free heap.

**Comment by pabloandresm** [comment]
This is how I discover the bug #52 https://github.com/mobizt/ESP32-Mail-Client/issues/52
Trying to send a picture of 90KB, only 64KB are OK.
I describe it in the issue #52 
I solved if changing the uint16_t to uint32_t in addAttachData(), add(), and some vectors, in ESP32_MailClient.h amd ESP32_MailClient.cpp.
After that It worked fine. I managed to send pictures of 150KB with no issues, arriving complete 100%.
